File: routes/admin_routes.py
"""
Admin routes: quản lý thực đơn, món ăn — Multi-tenant
"""
from flask import Blueprint, jsonify, request
from flask_login import login_required, current_user
from database import db, MonAn, ThucDon, DonVi, NguoiDung, get_user_don_vi_id
from datetime import datetime
import os, uuid

admin_bp = Blueprint('admin', __name__)

UPLOAD_FOLDER_ADMIN = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'static', 'uploads')


def admin_required(f):
    from functools import wraps
    @wraps(f)
    def decorated(*args, **kwargs):
        if not current_user.is_authenticated:
            return jsonify({'error': 'Chưa đăng nhập'}), 401
        if not current_user.is_admin:
            return jsonify({'error': 'Cần quyền quản trị viên'}), 403
        return f(*args, **kwargs)
    return decorated


# =========== QUẢN LÝ MÓN ĂN ===========

@admin_bp.route('/api/admin/dishes', methods=['GET'])
@login_required
def admin_get_dishes():
    """Lấy tất cả món ăn (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    dishes = MonAn.query.filter_by(don_vi_id=dvid).order_by(MonAn.loai, MonAn.ten).all()
    return jsonify([d.to_dict() for d in dishes])


@admin_bp.route('/api/admin/dishes', methods=['POST'])
@admin_required
def admin_create_dish():
    """Tạo món ăn mới (gán đơn vị)"""
    dvid = get_user_don_vi_id()
    if request.content_type and 'multipart/form-data' in request.content_type:
        data = request.form
        hinh_anh = None
        if 'hinh_anh' in request.files:
            file = request.files['hinh_anh']
            if file and file.filename:
                os.makedirs(UPLOAD_FOLDER_ADMIN, exist_ok=True)
                ext = file.filename.rsplit('.', 1)[-1].lower() if '.' in file.filename else 'jpg'
                fname = f"{uuid.uuid4().hex}.{ext}"
                file.save(os.path.join(UPLOAD_FOLDER_ADMIN, fname))
                hinh_anh = f'/static/uploads/{fname}'
    else:
        data = request.json
        hinh_anh = data.get('hinh_anh')

    mon = MonAn(
        ten=data['ten'],
        don_vi_id=dvid,
        mo_ta=data.get('mo_ta', ''),
        loai=data.get('loai', 'mon_chinh'),
        vung_mien=data.get('vung_mien', 'chung'),
        hinh_anh=hinh_anh,
        calo=float(data.get('calo', 0)),
        protein=float(data.get('protein', 0)),
        fat=float(data.get('fat', 0)),
        carbs=float(data.get('carbs', 0)),
        vitamin_a=float(data.get('vitamin_a', 0)),
        vitamin_c=float(data.get('vitamin_c', 0)),
        canxi=float(data.get('canxi', 0)),
        sat=float(data.get('sat', 0)),
    )
    db.session.add(mon)
    db.session.commit()
    return jsonify(mon.to_dict()), 201


@admin_bp.route('/api/admin/dishes/<int:dish_id>', methods=['PUT'])
@admin_required
def admin_update_dish(dish_id):
    """Cập nhật món ăn"""
    mon = MonAn.query.get_or_404(dish_id)
    if request.content_type and 'multipart/form-data' in request.content_type:
        data = request.form
        if 'hinh_anh' in request.files:
            file = request.files['hinh_anh']
            if file and file.filename:
                os.makedirs(UPLOAD_FOLDER_ADMIN, exist_ok=True)
                ext = file.filename.rsplit('.', 1)[-1].lower() if '.' in file.filename else 'jpg'
                fname = f"{uuid.uuid4().hex}.{ext}"
                file.save(os.path.join(UPLOAD_FOLDER_ADMIN, fname))
                mon.hinh_anh = f'/static/uploads/{fname}'
    else:
        data = request.json

    if 'ten' in data: mon.ten = data['ten']
    if 'mo_ta' in data: mon.mo_ta = data['mo_ta']
    if 'loai' in data: mon.loai = data['loai']
    if 'vung_mien' in data: mon.vung_mien = data['vung_mien']
    if 'calo' in data: mon.calo = float(data['calo'])
    if 'protein' in data: mon.protein = float(data['protein'])
    if 'fat' in data: mon.fat = float(data['fat'])
    if 'carbs' in data: mon.carbs = float(data['carbs'])
    if 'vitamin_a' in data: mon.vitamin_a = float(data['vitamin_a'])
    if 'vitamin_c' in data: mon.vitamin_c = float(data['vitamin_c'])
    if 'canxi' in data: mon.canxi = float(data['canxi'])
    if 'sat' in data: mon.sat = float(data['sat'])

    db.session.commit()
    return jsonify(mon.to_dict())


@admin_bp.route('/api/admin/dishes/<int:dish_id>', methods=['DELETE'])
@admin_required
def admin_delete_dish(dish_id):
    """Xóa món ăn"""
    mon = MonAn.query.get_or_404(dish_id)
    db.session.delete(mon)
    db.session.commit()
    return jsonify({'message': 'Da xoa mon an'})


# =========== QUẢN LÝ THỰC ĐƠN ===========

@admin_bp.route('/api/admin/menus', methods=['GET'])
@login_required
def admin_get_menus():
    """Lấy thực đơn (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    from datetime import date, timedelta
    start = date.today() - timedelta(days=7)
    end = date.today() + timedelta(days=14)
    menus = ThucDon.query.filter(
        ThucDon.ngay >= start, ThucDon.ngay <= end,
        ThucDon.don_vi_id == dvid
    ).order_by(ThucDon.ngay.desc(), ThucDon.bua).all()
    return jsonify([m.to_dict() for m in menus])


@admin_bp.route('/api/admin/menus', methods=['POST'])
@admin_required
def admin_create_menu():
    """Tạo thực đơn mới (gán đơn vị)"""
    dvid = get_user_don_vi_id()
    data = request.json
    ngay = datetime.strptime(data['ngay'], '%Y-%m-%d').date()

    # Xóa thực đơn cũ cùng ngày + bữa nếu có (cùng đơn vị)
    existing = ThucDon.query.filter_by(ngay=ngay, bua=data['bua'], don_vi_id=dvid).first()
    if existing:
        db.session.delete(existing)

    td = ThucDon(ngay=ngay, bua=data['bua'], don_vi_id=dvid, ghi_chu=data.get('ghi_chu', ''))
    if data.get('mon_an_ids'):
        mon_list = MonAn.query.filter(MonAn.id.in_(data['mon_an_ids']), MonAn.don_vi_id == dvid).all()
        td.mon_an_list = mon_list
    db.session.add(td)
    db.session.commit()
    return jsonify(td.to_dict()), 201


@admin_bp.route('/api/admin/menus/<int:menu_id>', methods=['PUT'])
@admin_required
def admin_update_menu(menu_id):
    """Cập nhật thực đơn"""
    dvid = get_user_don_vi_id()
    td = ThucDon.query.get_or_404(menu_id)
    data = request.json
    if 'mon_an_ids' in data:
        mon_list = MonAn.query.filter(MonAn.id.in_(data['mon_an_ids']), MonAn.don_vi_id == dvid).all()
        td.mon_an_list = mon_list
    if 'ghi_chu' in data:
        td.ghi_chu = data['ghi_chu']
    db.session.commit()
    return jsonify(td.to_dict())


@admin_bp.route('/api/admin/menus/<int:menu_id>', methods=['DELETE'])
@admin_required
def admin_delete_menu(menu_id):
    """Xóa thực đơn"""
    td = ThucDon.query.get_or_404(menu_id)
    db.session.delete(td)
    db.session.commit()
    return jsonify({'message': 'Da xoa thuc don'})


# =========== QUẢN LÝ NGƯỜI DÙNG ===========

@admin_bp.route('/api/admin/users')
@admin_required
def admin_get_users():
    """Lấy danh sách người dùng (cùng đơn vị)"""
    dvid = get_user_don_vi_id()
    users = NguoiDung.query.filter_by(don_vi_id=dvid).order_by(NguoiDung.ngay_tao.desc()).all()
    return jsonify([u.to_dict() for u in users])


@admin_bp.route('/api/admin/users/<int:uid>/toggle', methods=['POST'])
@admin_required
def admin_toggle_user(uid):
    u = NguoiDung.query.get_or_404(uid)
    u.kich_hoat = not u.kich_hoat
    db.session.commit()
    return jsonify(u.to_dict())


@admin_bp.route('/api/admin/users/<int:uid>/role', methods=['PUT'])
@admin_required
def admin_set_role(uid):
    u = NguoiDung.query.get_or_404(uid)
    data = request.json
    u.vai_tro = data.get('vai_tro', u.vai_tro)
    db.session.commit()
    return jsonify(u.to_dict())

File: routes/adminRoutes.js
/**
 * Admin routes - Quản trị
 */
const express = require('express');
const router = express.Router();
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const { v4: uuidv4 } = require('uuid');
const { db, getUserDonViId, thucDonToDict, nguoiDungToDict } = require('../database');

const adminRequired = (req, res, next) => {
  if (!req.session?.user) return res.status(401).json({ error: 'Chưa đăng nhập' });
  if (req.session.user.vai_tro !== 'admin') return res.status(403).json({ error: 'Cần quyền quản trị viên' });
  next();
};
const loginRequired = (req, res, next) => {
  if (!req.session?.user) return res.status(401).json({ error: 'Chua dang nhap' });
  next();
};

const uploadDir = path.join(__dirname, '..', 'static', 'uploads');
const storage = multer.diskStorage({
  destination: (req, file, cb) => { if (!fs.existsSync(uploadDir)) fs.mkdirSync(uploadDir, { recursive: true }); cb(null, uploadDir); },
  filename: (req, file, cb) => { const ext = path.extname(file.originalname).toLowerCase() || '.jpg'; cb(null, uuidv4().replace(/-/g, '') + ext); }
});
const upload = multer({ storage });

// ========= QUẢN LÝ MÓN ĂN =========
router.get('/api/admin/dishes', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const dishes = await db.all('SELECT * FROM mon_an WHERE don_vi_id = ? ORDER BY loai, ten', dvid);
    res.json(dishes);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/api/admin/dishes', adminRequired, upload.single('hinh_anh'), async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const data = req.body;
    const hinhAnh = req.file ? `/static/uploads/${req.file.filename}` : (data.hinh_anh || null);
    const r = await db.run(`INSERT INTO mon_an (ten, don_vi_id, mo_ta, loai, vung_mien, hinh_anh, calo, protein, fat, carbs, vitamin_a, vitamin_c, canxi, sat)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      data.ten, dvid, data.mo_ta || '', data.loai || 'mon_chinh', data.vung_mien || 'chung', hinhAnh,
      parseFloat(data.calo || 0), parseFloat(data.protein || 0), parseFloat(data.fat || 0), parseFloat(data.carbs || 0),
      parseFloat(data.vitamin_a || 0), parseFloat(data.vitamin_c || 0), parseFloat(data.canxi || 0), parseFloat(data.sat || 0)
    );
    res.status(201).json(await db.get('SELECT * FROM mon_an WHERE id = ?', r.lastInsertRowid));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.put('/api/admin/dishes/:dishId', adminRequired, upload.single('hinh_anh'), async (req, res) => {
  try {
    const mon = await db.get('SELECT * FROM mon_an WHERE id = ?', req.params.dishId);
    if (!mon) return res.status(404).json({ error: 'Not found' });
    const data = req.body;
    if (req.file) data.hinh_anh_path = `/static/uploads/${req.file.filename}`;
    const fields = ['ten', 'mo_ta', 'loai', 'vung_mien', 'calo', 'protein', 'fat', 'carbs', 'vitamin_a', 'vitamin_c', 'canxi', 'sat'];
    const updates = []; const values = [];
    fields.forEach(f => {
      if (data[f] !== undefined) {
        updates.push(`${f} = ?`);
        values.push(['calo','protein','fat','carbs','vitamin_a','vitamin_c','canxi','sat'].includes(f) ? parseFloat(data[f]) : data[f]);
      }
    });
    if (data.hinh_anh_path) { updates.push('hinh_anh = ?'); values.push(data.hinh_anh_path); }
    if (updates.length) {
      values.push(mon.id);
      await db.run(`UPDATE mon_an SET ${updates.join(', ')} WHERE id = ?`, ...values);
    }
    res.json(await db.get('SELECT * FROM mon_an WHERE id = ?', mon.id));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.delete('/api/admin/dishes/:dishId', adminRequired, async (req, res) => {
  try {
    const mon = await db.get('SELECT * FROM mon_an WHERE id = ?', req.params.dishId);
    if (!mon) return res.status(404).json({ error: 'Not found' });
    await db.run('DELETE FROM mon_an WHERE id = ?', mon.id);
    res.json({ message: 'Da xoa mon an' });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// ========= QUẢN LÝ THỰC ĐƠN =========
router.get('/api/admin/menus', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const today = new Date();
    const start = new Date(today); start.setDate(today.getDate() - 7);
    const end = new Date(today); end.setDate(today.getDate() + 14);
    const menus = await db.all('SELECT * FROM thuc_don WHERE ngay >= ? AND ngay <= ? AND don_vi_id = ? ORDER BY ngay DESC, bua',
      start.toISOString().split('T')[0], end.toISOString().split('T')[0], dvid);
    const result = [];
    for (const m of menus) { result.push(await thucDonToDict(m)); }
    res.json(result);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/api/admin/menus', adminRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const { ngay, bua, ghi_chu, mon_an_ids } = req.body;
    const existing = await db.get('SELECT id FROM thuc_don WHERE ngay = ? AND bua = ? AND don_vi_id = ?', ngay, bua, dvid);
    if (existing) {
      await db.run('DELETE FROM thuc_don_mon_an WHERE thuc_don_id = ?', existing.id);
      await db.run('DELETE FROM thuc_don WHERE id = ?', existing.id);
    }
    const r = await db.run('INSERT INTO thuc_don (ngay, bua, don_vi_id, ghi_chu) VALUES (?, ?, ?, ?)', ngay, bua, dvid, ghi_chu || '');
    if (mon_an_ids && mon_an_ids.length) {
      const ph = mon_an_ids.map(() => '?').join(',');
      const valid = await db.all(`SELECT id FROM mon_an WHERE id IN (${ph}) AND don_vi_id = ?`, ...mon_an_ids, dvid);
      for (const m of valid) { await db.run('INSERT INTO thuc_don_mon_an (thuc_don_id, mon_an_id) VALUES (?, ?)', r.lastInsertRowid, m.id); }
    }
    const td = await db.get('SELECT * FROM thuc_don WHERE id = ?', r.lastInsertRowid);
    res.status(201).json(await thucDonToDict(td));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.put('/api/admin/menus/:menuId', adminRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const td = await db.get('SELECT * FROM thuc_don WHERE id = ?', req.params.menuId);
    if (!td) return res.status(404).json({ error: 'Not found' });
    const { mon_an_ids, ghi_chu } = req.body;
    if (mon_an_ids) {
      await db.run('DELETE FROM thuc_don_mon_an WHERE thuc_don_id = ?', td.id);
      const ph = mon_an_ids.map(() => '?').join(',');
      const valid = await db.all(`SELECT id FROM mon_an WHERE id IN (${ph}) AND don_vi_id = ?`, ...mon_an_ids, dvid);
      for (const m of valid) { await db.run('INSERT INTO thuc_don_mon_an (thuc_don_id, mon_an_id) VALUES (?, ?)', td.id, m.id); }
    }
    if (ghi_chu !== undefined) await db.run('UPDATE thuc_don SET ghi_chu = ? WHERE id = ?', ghi_chu, td.id);
    res.json(await thucDonToDict(await db.get('SELECT * FROM thuc_don WHERE id = ?', td.id)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.delete('/api/admin/menus/:menuId', adminRequired, async (req, res) => {
  try {
    const td = await db.get('SELECT * FROM thuc_don WHERE id = ?', req.params.menuId);
    if (!td) return res.status(404).json({ error: 'Not found' });
    await db.run('DELETE FROM thuc_don_mon_an WHERE thuc_don_id = ?', td.id);
    await db.run('DELETE FROM thuc_don WHERE id = ?', td.id);
    res.json({ message: 'Da xoa thuc don' });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// ========= QUẢN LÝ NGƯỜI DÙNG =========
router.get('/api/admin/users', adminRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const users = await db.all('SELECT * FROM nguoi_dung WHERE don_vi_id = ? ORDER BY ngay_tao DESC', dvid);
    res.json(users.map(nguoiDungToDict));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/api/admin/users/:uid/toggle', adminRequired, async (req, res) => {
  try {
    const u = await db.get('SELECT * FROM nguoi_dung WHERE id = ?', req.params.uid);
    if (!u) return res.status(404).json({ error: 'Not found' });
    await db.run('UPDATE nguoi_dung SET kich_hoat = ? WHERE id = ?', u.kich_hoat ? 0 : 1, u.id);
    res.json(nguoiDungToDict(await db.get('SELECT * FROM nguoi_dung WHERE id = ?', u.id)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.put('/api/admin/users/:uid/role', adminRequired, async (req, res) => {
  try {
    const u = await db.get('SELECT * FROM nguoi_dung WHERE id = ?', req.params.uid);
    if (!u) return res.status(404).json({ error: 'Not found' });
    await db.run('UPDATE nguoi_dung SET vai_tro = ? WHERE id = ?', req.body.vai_tro || u.vai_tro, u.id);
    res.json(nguoiDungToDict(await db.get('SELECT * FROM nguoi_dung WHERE id = ?', u.id)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
